ex-4-quicksort/ex-4-2-recursive-list-count.js

```diff
@@ -0,0 +1,13 @@
+//Exercise 4.2
+//Count a list of items with recusion
+
+function count( arr ) {
+    if ( arr.length <= 0 ) {
+        return 0;
+    }
+
+    return 1 + count( [...arr.splice(1, arr.length)] );
```

Comment:
I noticed you used `splice` here and `slice` for `sum`. What are the differences between these two? Are there hidden/unexpected implications for the consumer of your `count`?

Reply:
Thanks for catching this!  I didn't have a good reason to use `splice` here.  I think I meant to use `slice` and even passed it the parameters I'd expect to see in `slice`.  I'm lucky it even worked.  

>What are the differences between these two? Are there hidden/unexpected implications for the consumer of your count?

`Splice` and `slice` do different things.  Splice replaces elements in an array, mutating that array in the process.  `Slice` creates a shallow copy of an array, leaving the original array untouched.  Using `splice` made the `count()` function impure, causing the side effect of mutating the original array.  Using `slice`, instead, makes this a pure function.  

Updated to use `slice`: https://github.com/mdmoore/grokking-algorithms/pull/6/commits/3f8d4df7be25f1ba5882bce69789e019c9d6ed8a

Reply:
>  Using splice made the count() function impure, causing the side effect of mutating the original array. Using slice, instead, makes this a pure function.

100%, right on.